initial upload of yum_versionlock

gyptazy opened the pull request:
Adds support to lock packages in their version to prevent these from updating by yum.

##### SUMMARY
This module adds support to lock packages in their version to prevent these from updating by yum.

Fixes #13086 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
- yum_versionlock

##### ANSIBLE VERSION
```
ansible 2.4.2.0
```

##### ADDITIONAL INFORMATION
An example run in local playbook

```
TASK [version_lock : lock nginx to prevent updates] **********************************************************************************************************
changed: [localhost]

TASK [version_lock : update nginx] ***************************************************************************************************************************
ok: [localhost]

TASK [version_lock : unlock nginx from updates] **************************************************************************************************************
changed: [localhost]

TASK [version_lock : update nginx] ***************************************************************************************************************************
changed: [localhost]
```

florianpaulhoberg committed:
initial upload of yum_versionlock

adds support for to lock packages in their version to prevent these from updating

florianpaulhoberg committed:
Update yum_versionlock.py

fixed checkmode var

florianpaulhoberg committed:
Update yum_versionlock.py

fixed trailing whitespaces

florianpaulhoberg committed:
code cleanup

- code fixes for pep
- code fixes for ansible-test

florianpaulhoberg committed:
fixed ansible-doc

 - fixed ansible-doc

florianpaulhoberg committed:
added choices to argument_spec

- added choices to argument_spec
- fixed docu

florianpaulhoberg committed:
added missing return block

florianpaulhoberg committed:
Update yum_versionlock.py

 codefix for test (fixes: E713)

florianpaulhoberg committed:
Update yum_versionlock.py

fixed whitespaces in return

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+---
+module: yum_versionlock
+version_added: 2.7
+short_description: Locks/prevents an installed package from beeing updates by (yum) package mamanger.
```
Add `Locks/Unlocks` - as both operations are allowed

Akasurde replied:
s/beeing/being/

Akasurde replied:
`s/mamanger/manager/`

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+options:
+  package:
+    description:
+      - A package name like C(httpd).
```
Can we have a list of packages which can be locked/unlocked? just like yum module does.

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+      - A package name like C(httpd).
+  state:
+    description:
+      - Whether to lock (C(present) or unlock (C(absent) a package.
```
(C(present) -> C(present)
(C(absent) -> C(absent)

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+- yum
+- yum-versionlock
+author:
+    - Florian Paul Hoberg <florian.hoberg@credativ.de>
```
add github handle too.

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+import os.path
+from ansible.module_utils.basic import AnsibleModule
+
+YUM_BINARY = "/bin/yum"
```
use this instead

```
module.get_bin_path('yum', required=True)
```

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+
+def get_state_yum_versionlock():
+    """ Check for yum plugin dependency """
+    state = os.path.exists("/etc/yum/pluginconf.d/versionlock.conf")
```
Are there any other way to check if version lock plugin is installed ?

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+    """ Get an overview of all packages on yum versionlock """
+    rc_code, out, err = module.run_command("%s -q versionlock list"
+                                           % (YUM_BINARY))
+    if rc_code is 0:
```
```
rc_code == 0
```

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+    """ start main program to add/remove a package to yum versionlock"""
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(required=True, type='str', choices=['present', 'absent']),
```
state can be optional with default value as 'present'. Also, default argument type is str so no need to specify explicitly.

Akasurde commented on lib/ansible/modules/packaging/os/yum_versionlock.py:
```diff
@@ -0,0 +1,148 @@
+version_added: 2.7
+short_description: Locks/prevents an installed package from beeing updates by (yum) package mamanger.
+description:
+     - This module adds installed packages to yum versionlock to prevent it from beeing updated.
```
s/beeing/being/

Akasurde reviewed (changes requested):
Could you please check inline comments and also add testcase for this module ? thanks.

florianpaulhoberg committed:
Update yum_versionlock.py

- fixed typos
- get yum binary from module.get_bin_path
- check for installed dependency "yum-plugin-versionlock" with rpm module

florianpaulhoberg committed:
Update yum_versionlock.py

- removed yum import
- added github handle

florianpaulhoberg committed:
Update yum_versionlock.py

- added missing blank lines

florianpaulhoberg committed:
Update yum_versionlock.py

florianpaulhoberg commented:
ok thanks, unit test file will be created soon.

florianpaulhoberg committed:
added support to define multiple packages

avarins commented:
What is the status?
Does a ready_for_review have to be attached?

corbinstuard commented:
This looks good, can it be reviewed?

a-nldisr commented:
@Akasurde any plans to check this, would be great to have! :)

riemers commented:
Want this the Ansible way, would need this too. But since its 2 years old, i don't think i should wait on it. (which is a shame) i could even look into it if that would be the default reply (as far as my knowledge goes)

Akasurde commented:
@riemers @florianpaulhoberg All community modules are migrated to https://github.com/ansible-collections/community.general repo. Could you please migrate this PR to above-mentioned repo. I am sorry that I missed previous ping messages. Thanks.

riemers commented:
I moved it over as is, we shall see how far that gets us :)

Akasurde commented:
@riemers Thanks for migrating this module to new repo.

Migrated to https://github.com/ansible-collections/community.general/pull/324
